File: gateway/src/services/profileService.js
const { logger } = require('../config/logger');

class ProfileService {
  constructor() {
    // Load profiles from configuration
    this.profiles = this._loadProfiles();
  }

  /**
   * Get all available profiles
   */
  async getAllProfiles() {
    return this.profiles.filter(p => p.isActive);
  }

  /**
   * Get a specific profile by ID
   */
  async getProfile(profileId) {
    return this.profiles.find(p => p.id === profileId);
  }

  /**
   * Get capabilities for a profile
   */
  async getProfileCapabilities(profileId) {
    const profile = await this.getProfile(profileId);
    return profile ? profile.capabilities : [];
  }

  /**
   * Update profile configuration
   */
  async updateProfile(profileId, updates) {
    const index = this.profiles.findIndex(p => p.id === profileId);
    if (index !== -1) {
      this.profiles[index] = { ...this.profiles[index], ...updates };
      logger.info('Profile updated', { profileId });
      return this.profiles[index];
    }
    return null;
  }

  /**
   * Load profiles from configuration
   */
  _loadProfiles() {
    // TODO: Load from database or configuration file
    return [
      {
        id: 'weather',
        name: 'Weather Assistant',
        description: 'Get weather information and forecasts',
        icon: '🌤️',
        version: '1.0.0',
        capabilities: ['weather_current', 'weather_forecast', 'weather_alerts'],
        configuration: {
          api_endpoint: '/weather',
          cache_duration: 300,
        },
        isActive: true,
        mcpServerUrl: 'http://localhost:3001/mcp/weather',
      },
      {
        id: 'calendar',
        name: 'Calendar Manager',
        description: 'Manage your calendar events and reminders',
        icon: '📅',
        version: '1.0.0',
        capabilities: ['calendar_read', 'calendar_write', 'reminders'],
        configuration: {
          api_endpoint: '/calendar',
          sync_enabled: true,
        },
        isActive: true,
        mcpServerUrl: 'http://localhost:3001/mcp/calendar',
      },
      {
        id: 'notes',
        name: 'Notes Assistant',
        description: 'Create and manage your notes',
        icon: '📝',
        version: '1.0.0',
        capabilities: ['notes_create', 'notes_read', 'notes_search'],
        configuration: {
          api_endpoint: '/notes',
          markdown_enabled: true,
        },
        isActive: true,
        mcpServerUrl: 'http://localhost:3001/mcp/notes',
      },
      {
        id: 'tasks',
        name: 'Task Manager',
        description: 'Organize and track your tasks',
        icon: '✅',
        version: '1.0.0',
        capabilities: ['tasks_create', 'tasks_update', 'tasks_complete'],
        configuration: {
          api_endpoint: '/tasks',
          priority_levels: ['low', 'medium', 'high'],
        },
        isActive: true,
        mcpServerUrl: 'http://localhost:3001/mcp/tasks',
      },
    ];
  }
}

module.exports = { ProfileService };

File: mobile_app/lib/services/profile_service.dart
import 'package:flutter/foundation.dart';
import '../models/mcp_profile.dart';
import 'dart:convert';

/// Manages MCP profiles and their configurations
class ProfileService extends ChangeNotifier {
  List<MCPProfile> _profiles = [];
  MCPProfile? _selectedProfile;
  bool _isLoading = false;

  List<MCPProfile> get profiles => List.unmodifiable(_profiles);
  MCPProfile? get selectedProfile => _selectedProfile;
  bool get isLoading => _isLoading;

  ProfileService() {
    _loadProfiles();
  }

  /// Load available MCP profiles
  Future<void> _loadProfiles() async {
    _isLoading = true;
    notifyListeners();

    try {
      // TODO: Load profiles from MCP Gateway or local storage
      // For now, create mock profiles
      _profiles = _getMockProfiles();
      
      if (_profiles.isNotEmpty && _selectedProfile == null) {
        _selectedProfile = _profiles.first;
      }
    } catch (e) {
      debugPrint('Error loading profiles: $e');
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  /// Get mock profiles for demonstration
  List<MCPProfile> _getMockProfiles() {
    return [
      MCPProfile(
        id: 'weather',
        name: 'Weather Assistant',
        description: 'Get weather information and forecasts',
        icon: '🌤️',
        version: '1.0.0',
        capabilities: ['weather_current', 'weather_forecast', 'weather_alerts'],
        configuration: {
          'api_endpoint': '/weather',
          'cache_duration': 300,
        },
      ),
      MCPProfile(
        id: 'calendar',
        name: 'Calendar Manager',
        description: 'Manage your calendar events and reminders',
        icon: '📅',
        version: '1.0.0',
        capabilities: ['calendar_read', 'calendar_write', 'reminders'],
        configuration: {
          'api_endpoint': '/calendar',
          'sync_enabled': true,
        },
      ),
      MCPProfile(
        id: 'notes',
        name: 'Notes Assistant',
        description: 'Create and manage your notes',
        icon: '📝',
        version: '1.0.0',
        capabilities: ['notes_create', 'notes_read', 'notes_search'],
        configuration: {
          'api_endpoint': '/notes',
          'markdown_enabled': true,
        },
      ),
      MCPProfile(
        id: 'tasks',
        name: 'Task Manager',
        description: 'Organize and track your tasks',
        icon: '✅',
        version: '1.0.0',
        capabilities: ['tasks_create', 'tasks_update', 'tasks_complete'],
        configuration: {
          'api_endpoint': '/tasks',
          'priority_levels': ['low', 'medium', 'high'],
        },
      ),
    ];
  }

  /// Select a profile
  void selectProfile(MCPProfile profile) {
    _selectedProfile = profile;
    notifyListeners();
  }

  /// Refresh profiles from the server
  Future<void> refreshProfiles() async {
    await _loadProfiles();
  }

  /// Update profile configuration
  Future<bool> updateProfile(MCPProfile profile) async {
    try {
      final index = _profiles.indexWhere((p) => p.id == profile.id);
      if (index != -1) {
        _profiles[index] = profile;
        if (_selectedProfile?.id == profile.id) {
          _selectedProfile = profile;
        }
        notifyListeners();
        return true;
      }
      return false;
    } catch (e) {
      debugPrint('Error updating profile: $e');
      return false;
    }
  }

  /// Toggle profile active status
  Future<void> toggleProfileStatus(String profileId) async {
    final profile = _profiles.firstWhere((p) => p.id == profileId);
    final updatedProfile = profile.copyWith(isActive: !profile.isActive);
    await updateProfile(updatedProfile);
  }
}
